exec/cnex/number.c

```diff
@@ -77,79 +91,89 @@ void number_toString(Number x, char *buf, size_t len)
             int count = 2;
-            // Start out by setting [r]esult to "0."
-            memcpy(r, "0.", count);
+            // Start out by setting buf to "0."
+            memcpy(buf, "0.", count);
             // Now, we copy the necessary number of zeros after the decimal point...
             for (int i = 0; i < -exponent-slen;i++, count++) {
-                r[count] = '0';
+                buf[count] = '0';
             }
-            // Then copy the remainder of the mantissa to our [r]esult.
-            memcpy(&r[count], mantissa, slen);
-            // Ensure that [r]esult is null terminated.  I'm not 100% convinced this is necessary.
+            // Then copy the remainder of the mantissa to the supplied buf
+            memcpy(&buf[count], mantissa, slen);
+            // Ensure that buf is null terminated.  I'm not 100% convinced this is necessary.
             slen += count;
-            r[slen] = '\0';
+            buf[slen] = '\0';
         } else {
             // Our exponent is positive, therefore we need to add the decimal point on the right of the number.
             exponent += last_significant_digit;
             if (last_significant_digit >= 1) {
-                // Our mantissa is at least 2 digits long, so copy the first digit to [r]esult
-                r[0] = mantissa[0];
+                // Our mantissa is at least 2 digits long, so copy the first digit to buf
+                buf[0] = mantissa[0];
                 // add the decimal point
-                r[1] = '.';
-                // copy the remainder of the mantissa to [r]esult
-                memcpy(&r[2], &mantissa[1], last_significant_digit);
+                buf[1] = '.';
+                // copy the remainder of the mantissa to buf
+                memcpy(&buf[2], &mantissa[1], last_significant_digit);
                 slen++;
             } else {
                 // The mantissa is only one significant digit, so copy that to the output, then apply the exponent value.
-                memcpy(r, mantissa, last_significant_digit+1);
+                memcpy(buf, mantissa, last_significant_digit+1);
             }
             // Now, create the actual exponent value.
             char exp[36];
             int explen = sprintf(exp, "e%d", exponent);
-            memcpy(&r[slen], exp, explen+1); // Include the terminating NULL char given to us by sprintf()
+            memcpy(&buf[slen], exp, explen+1); // Include the terminating NULL char given to us by sprintf()
             slen += explen;
         }
     } else {
         // Exponent is zero, or just copy the mantissa, up to the last significant digit.
-        memcpy(r, mantissa, last_significant_digit+1); // Include terminating NULL char.
+        memcpy(buf, mantissa, last_significant_digit+1); // Include terminating NULL char.
+    }
+    if ((size_t)(slen + 1) > len) {
+        // The caller didn't supply us with enough buffer space to hold the resulting number.
```

Comment:
It seems a bit late to double check the buffer size here, after having already put the data in the buffer. In what situation would this check be executed?